HandsOnProject/app/src/main/java/com/example/handsonproject/MainActivity.kt

```diff
@@ -1,11 +1,144 @@
+    }
+
+
+    fun onTapB1(view: View){
+        with(binding) { editText1.setText(editText1.getText().toString() + "1") }
+    }
+    fun onTapB2(view: View){
+        with(binding) { editText1.setText(editText1.getText().toString() + "2") }
+    }
+    fun onTapB3(view: View){
+        with(binding) { editText1.setText(editText1.getText().toString() + "3") }
+    }
+    fun onTapB4(view: View){
+        with(binding) { editText1.setText(editText1.getText().toString() + "4") }
+    }
+    fun onTapB5(view: View){
+        with(binding) { editText1.setText(editText1.getText().toString() + "5") }
+    }
+    fun onTapB6(view: View){
+        with(binding) { editText1.setText(editText1.getText().toString() + "6") }
+    }
+    fun onTapB7(view: View){
+        with(binding) { editText1.setText(editText1.getText().toString() + "7") }
+    }
+    fun onTapB8(view: View){
+        with(binding) { editText1.setText(editText1.getText().toString() + "8") }
+    }
+    fun onTapB9(view: View){
+        with(binding) { editText1.setText(editText1.getText().toString() + "9") }
+    }
+    fun onTapB0(view: View){
+        with(binding) { editText1.setText(editText1.getText().toString() + "0") }
+    }
+    fun onTapBEqual(view: View){
+        with(binding){
+            if (editText1.text.toString().equals("")&&textView.text.toString().equals("")) return
+            if(op==1){
+                val res=editText1.text.toString().toDouble()+textView.text.toString().toDouble()
+                editText1.setText(res.toString())
+                textView.setText("")
+            }
+            if(op==2){
+                val res=textView.text.toString().toDouble()-editText1.text.toString().toDouble()
+                editText1.setText(res.toString())
+                textView.setText("")
+            }
+            if(op==3){
+                val res=editText1.text.toString().toDouble()*textView.text.toString().toDouble()
+                editText1.setText(res.toString())
+                textView.setText("")
+            }
+            if(op==4){
+                val res=textView.text.toString().toDouble()/editText1.text.toString().toDouble()
+                editText1.setText(res.toString())
+                textView.setText("")
+            }
+            op=0
+        }
+    }
+    fun onTapBPlus(view: View){
+        with(binding) {
+            if (op==0) {
+                textView.setText(editText1.getText().toString())
+                op=1
+            }
+            else {
+                onTapBEqual(view)
+                onTapBPlus(view)
+            }
+            editText1.setText("")
+        }
+    }
+    fun onTapBMinus(view: View){
+        with(binding) {
+            if (op==0) {
+                textView.setText(editText1.getText().toString())
+                op=2
+            }
+            else {
+                onTapBEqual(view)
+                onTapBMinus(view)
+            }
+            editText1.setText("")
+        }
+    }
```

Comment:
@shovian I found that the app crashes with the following procedure:

1. Input a number
2. Click `-` button
3. Click `-` button again

Please check

Reply:
@shovian Am I right to assume that you also fixed the crash? If so, please explain the cause of the crash and how you fixed it. Thanks!

Reply:
@shovian 

> So, what I did just put a case when the input is blank then it will return.

OK let's accept this as a working solution for now.

> I'm not quite sure about the cause of the crash

Still, I think it would be better to know what actually happened. Revert it into a state where it is crashing, and then look at the `catlog`. Usually it will give you the exception (what happened) and a stack trace (where it happened) including even line numbers. 

Try to figure out the cause. If you're having trouble, Yamauchi-san or I can assist you.

Reply:
okay then!

Reply:
it turns out because the null handler in onTapBEqual return when one of the stash or the input is null, the state of the operation didn't turn into MAIN_STATUS, and it makes the call on onTapB<Operation> get into the same exact condition and doing infinite loops

Reply:
@shovian 

Thanks for the analysis.